src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedNewTests.cs

```diff
@@ -0,0 +1,928 @@
+                );
+        }
+
+        [Fact]
+        public void TestInaccessibleConstructor()
+        {
+            var comp = CreateCompilation(@"
+class C { private C() {} }
+
+class X {
+    public static void Main() {
+        C x = new();
+    }
+}
+", options: TestOptions.ReleaseExe).VerifyDiagnostics(
+                // (6,15): error CS0122: 'C.C()' is inaccessible due to its protection level
+                //         C x = new();
+                Diagnostic(ErrorCode.ERR_BadAccess, "new()").WithArguments("C.C()").WithLocation(6, 15)
+                );
+        }
+
+        [Fact]
+        public void TestBadArgs()
+        {
+            var comp = CreateCompilation(@"
+class C { private C() {} }
+
+class X {
+    public static void Main() {
+        C x = new(1);
+    }
+}
+", options: TestOptions.ReleaseExe).VerifyDiagnostics(
+                // (6,15): error CS1729: 'C' does not contain a constructor that takes 1 arguments
+                //         C x = new(1);
+                Diagnostic(ErrorCode.ERR_BadCtorArgCount, "new(1)").WithArguments("C", "1").WithLocation(6, 15)
+                );
+        }
+
+        [Fact]
+        public void TestNested()
+        {
+            var comp = CreateCompilation(@"
+using System;
+class C { 
+    public C(C a, C b) => Console.Write(3); 
+    public C(int i) => Console.Write(i);
+}
+
+class X {
+    public static void Main() {
+        C x = new(new(1), new(2));
+    }
+}
+", options: TestOptions.ReleaseExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "123");
+        }
+
+        [Fact]
+        public void TestBestType_TypeArg()
+        {
+            var comp = CreateCompilation(@"
+using System;
+
+class X {
+    public static void M<T>(T t1, T t2) {
+        Console.Write($""{t1}{t2}"");
+    }
+
+    public static void Main() {
+        M(new X(), new());
+    }
+}
+", options: TestOptions.ReleaseExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "XX");
+        }
+
+        [Fact]
+        public void TestBestType_ArrayInit()
+        {
+            var comp = CreateCompilation(@"
+using System;
+
+class X {
+    public static void Main() {
+        var arr = new[] {new X(), new()};
+        Console.Write($""{arr[0]}{arr[1]}"");
+    }
+}
+", options: TestOptions.ReleaseExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "XX");
+        }
+
+        [Fact]
+        public void TestBestType_NullCoalescing()
+        {
+            var comp = CreateCompilation(@"
+using System;
+
+class X {
+    public static void Main() {
+        X x = null;
+        Console.Write((x ?? new()));
+    }
+}
+", options: TestOptions.ReleaseExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "X");
+        }
+
+        [Fact]
+        public void TestBestType_Lambda()
```

Comment:
>TestBestType_Lambda [](start = 20, length = 19)

Consider adding a variant of this test where all returns in the lambda use `new()`, so the type inference should fail.